core/domain/email_services.py

```diff
@@ -110,7 +112,10 @@ def send_mail(
             send_email_to_recipients() function returned False
             (signifying API returned bad status code).
     """
-    if not feconf.CAN_SEND_EMAILS:
+    recipient_email_in_allowlist = recipient_email in (
+        feconf.EMAIL_RECIPIENT_ALLOWLIST_FOR_LOGGING)
+
+    if not feconf.CAN_SEND_EMAILS and not recipient_email_in_allowlist:
```

Comment:
As part of this PR can we make sure to remove `CAN_SEND_EMAILS` from the other parts of the codebase, we have it in many other places and that means many of the emails won't get here to even be logged.

Reply:
Hmm, shouldn't we update it everywhere then: i.e. check for these two conditions everywhere. Not sure why do we need to remove it completely?

Reply:
I think these were just added everywhere to make stuff easier and we discussed previously that we want to only use it in this place. Since all emails should end here anyway.

Reply:
Just to check, is it definitely the case that this check can't be bypassed without raising a flag for reviewers? If so, then it should be safe to only have the check here. But we might need to make sure to add notes / comments etc. to warn folks away from calling the underlying methods directly, or make those private in some way.

Reply:
@seanlip Sorry, I don't understand what you're talking about.

Reply:
Suppose a future developer writes code that bypasses the "can send emails" check by calling the underlying code which this method calls to send emails.

Are reviewers likely to catch this and request that it be fixed?

Reply:
Ah, you mean the checks that are in functions that we call from this function? If that sp then I would be for removing the check from here and moving it as further as we can (directly before the call that would fail on the backup or test server). 

I'm mostly talking about the checks that are above this and are called before this check. These should be removed in favor of this check.

Reply:
@vojtechjelinek I have looked into this and there a couple of places where `feconf.CAN_SEND_EMAILS` or vars like `feconf.CAN_SEND_EDITOR_ROLE_EMAILS` are used before calling the actual function to send emails. Due to this, an exception won't be raised or error won't be logged if sending emails is not allowed since the error raising part is done in the function in email manager.

Now, if we remove this check from other libraries, we will see errors being logged or exceptions being raised where they were not in the first place.

I think a better way would be to introduce a new function `can_send_or_log_emails` which will return the valid recipient email id list to which emails can be sent or logged and use this instead of feconf checks everywhere. WDYT?

Reply:
I have added the function in [this commit](https://github.com/oppia/oppia/pull/15408/commits/0a64ce705b2631e0184e4e5c012e731feb0a6bef), let me know if we should proceed with this approach.

Reply:
Sorry, I don't see what might be the problem, especially this paragraph:

> Now, if we remove this check from other libraries, we will see errors being logged or exceptions being raised where they were not in the first place.

If you mean that the issue might be that we will start receiving new errors from places in our code where we didn't before, I don't see a big issue as I only expect these errors in dev mode and not-prod servers, which should be fine.